### DISCUSSION TITLE
Start an workflow from a definition and grab the output

### DISCUSSION THREAD

**Discussion by robverhoef-united** [elsa-workflows/elsa-core#5052] [Q&A]
Hello,

The simplest way of starting a workflow using a DefinitionID i found is:

`var w = new StartWorkflowRuntimeOptions(); `
`w.Input = d!;`
`var result = await _workflowRuntime.StartWorkflowAsync(workflowDefinitionId, w);`

Somehow i can't figure out how to get hold of the output of the started (and finished) workflow from the `result` var.

Al the examples i found all use the `IWorkflowRunner` but that does not have a function which i can supply the `WorkflowDefinitionId `to.

Tried this, but does not seem to start anything, it seems that the `workflow` var is initiated but empty.

`var workflowBuilderFactory = context.GetRequiredService<IWorkflowBuilderFactory>();`
`var workflowBuilder = workflowBuilderFactory.CreateBuilder();`
`workflowBuilder.DefinitionId = workflowDefinitionId;`
`workflowBuilder.Version = 1;`
`var workflow = await workflowBuilder.BuildWorkflowAsync();`
`var workflowRunner = context.GetRequiredService<IWorkflowRunner>();`
`var result = await workflowRunner.RunAsync(workflow, new RunWorkflowOptions { Input = d! });`

I define my workflows using  code like this using the DefinitionId below in the functions above.

`public class FFProbeWorkflow : WorkflowBase`
`{`
`    protected override void Build(IWorkflowBuilder builder)`
`    {`
`          builder.DefinitionId = "FFProbeWorkflow";`
`           builder.Name = "FFProbe a file and store result";`
`           builder.Version = 1;`
`           builder.Root = new Flowchart`
`           {`
`               Variables = { ........},`
`               Activities =`
`               {`
`                   ...............`
`               },`
`               Connections =`
`               {`
`                   .................`
`               }`
`           };`
`      }`
`}`

Any help is very much appreciated!

Regards, Rob

**Comment by glime-ai** [comment]
Hello Rob,

To obtain the output of a workflow that you've started using the `IWorkflowRuntime` service, you can access the output from the `WorkflowExecutionContext` object's `Output` property, as shown in the first context document you provided. Here's a simplified version of how you can do it:

```csharp
var input = new Dictionary<string, object>
{
    ["YourInputKey"] = yourInputValue
};

var definitionId = "FFProbeWorkflow"; // Your workflow definition ID
var startWorkflowOptions = new StartWorkflowRuntimeOptions(input: input);
var result = await _workflowRuntime.StartWorkflowAsync(definitionId, startWorkflowOptions);

// Assuming your workflow has an output with the key "YourOutputKey"
var output = result.WorkflowExecutionContext.Output["YourOutputKey"];
```

In this code snippet, replace `"YourInputKey"` and `yourInputValue` with the actual input key and value you want to pass to the workflow. Similarly, replace `"YourOutputKey"` with the key that your workflow uses to store its output.

If you're not seeing any output, make sure that your workflow is correctly setting the output in its activities. Each activity has access to the `WorkflowExecutionContext` object via the `ActivityExecutionContext` parameter that is passed to the `ExecuteAsync` method, and you can set the output using this context.

If you're still having trouble, please provide more details about the workflow's activities and how they're supposed to set the output, and I can give you more specific advice.

**Reply by robverhoef-united** [reply]
Hello,

Thanks for the prompt answer, the result i get is of type WorkFlowExecutionResult and does not seem to contain a WorkflowExecutionContext. ??

I just found another solution that does the trick though. i use:
`var state = await _workflowRuntime.ExportWorkflowStateAsync(result.WorkflowInstanceId);`
This is the same state that the IWorkflowRunner returns and holds my output.

Thanks, Rob
